synapse/rest/client/login.py

```diff
@@ -342,6 +343,16 @@ async def _complete_login(
             user_id = canonical_uid
 
         device_id = login_submission.get("device_id")
+
+        # Check that device_id is not greater than a reasonable 512B
+        device_id_size = sys.getsizeof(device_id)
+        if device_id_size > 512:
+            raise LoginError(
+                400,
+                "Device_id cannot be greater than 512B.",
+                errcode=Codes.INVALID_PARAM,
+            )
+
```

Comment:
I have two minor nitpicks which you should might want to ignore for the purposes of this review, but which might be of interest.

Nitpicking: this is asserting that (C)Python's internal representation of the string is no larger than 512 bytes. Postgres likey has a different representation with a different size, but the two are almost certainly close enough so that there we'll still be within the 8KB limit.

Another nitpick: future versions of CPython might choose to use a different string representation where the same string is smaller or larger in-memory. On my system, Python 3.10 uses 49 bytes for the empty string, whereas Python 2 uses 37 bytes. 

```shell
$ python2 --version; python2 -c "import sys; print(sys.getsizeof(''))"
Python 2.7.18
37

$ python3 --version; python3 -c "import sys; print(sys.getsizeof(''))"
Python 3.10.4
49
```

If this changes under our feet then we might have device IDs that are rejected after a homeserver runner upgrades their Python version.

Reply:
I think we usually limit by string length (number of characters)?

Reply:
Very useful nitpicks! I have changed it to limit by size, I appreciate the explanation.